man/getdata.3
.\" getdata.3.  The getdata man page.
.\"
.\" (C) 2008, 2009, 2010 D. V. Wiebe
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\"
.\" This file is part of the GetData project.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.2 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, with no Front-Cover Texts, and with no Back-Cover
.\" Texts.  A copy of the license is included in the `COPYING.DOC' file
.\" as part of this distribution.
.\"
.TH getdata 3 "9 February 2010" "Version 0.6.1" "GETDATA"
.SH NAME
getdata \(em retrieve data from a dirfile database
.SH SYNOPSIS
.B #include <getdata.h>
.HP
.nh
.ad l
.BI "size_t getdata(DIRFILE *" dirfile ", const char *" field_code ", off_t"
.IB first_frame ", off_t " first_sample ", size_t " num_frames ", size_t"
.IB num_samples ", gd_type_t " return_type ", void *" data_out );
.hy
.ad n
.SH DESCRIPTION
The
.BR getdata ()
function queries a dirfile(5) database specified by
.I dirfile
for the field
.IR field_code .
It fetches
.I num_frames
frames plus
.I num_samples
samples from this field, starting 
.I first_sample
samples past frame
.IR first_frame . 
The data is converted to the data type specified by
.IR return_type ,
and stored in the user-supplied buffer
.IR data_out .

The 
.I field_code
may contain one of the representation suffixes listed in dirfile-format(5).
If it does, getdata will compute the appropriate complex norm before returning
the data.

The 
.I dirfile
argument must point to a valid DIRFILE object previously created by a call to
.BR dirfile_open (3).
The argument
.I data_out
must point to a valid memory location of sufficient size to hold all data
requested.

The first sample returned will be
.RS
.IR first_frame " * " samples_per_frame " + " first_sample
.RE
as measured from the start of the dirfile, where
.I samples_per_frame
is the number of samples per frame as returned by
.BR get_spf (3).
The number of samples fetched is, similarly,
.RS
.IR num_frames " * " samples_per_frame " + " num_samples .
.RE
Although calling
.BR getdata ()
using both samples and frames is possible, the function is typically called
with either
.IR num_samples " and " first_sample ,
or
.IR num_frames " and " first_frames ,
equal to zero.

The 
.I return_type
argument should be one of the following symbols, which indicates the desired
return type of the data:
.RS
.TP 11
.I GD_UINT8
unsigned 8-bit integer
.TP
.I GD_INT8
signed (two's complement) 8-bit integer
.TP
.I GD_UINT16
unsigned 16-bit integer
.TP
.I GD_INT16
signed (two's complement) 16-bit integer
.TP
.I GD_UINT32
unsigned 32-bit integer
.TP
.I GD_INT32
signed (two's complement) 32-bit integer
.TP
.I GD_UINT64
unsigned 64-bit integer
.TP
.I GD_INT64
signed (two's complement) 64-bit integer
.TP
.IR GD_FLOAT32
IEEE-754 standard 32-bit single precision floating point number
.TP
.IR GD_FLOAT64
IEEE-754 standard 64-bit double precision floating point number
.TP
.IR GD_COMPLEX64
C99-conformant 64-bit single precision complex number
.TP
.IR GD_COMPLEX128
C99-conformant 128-bit double precision complex number
.TP
.I GD_NULL
the null type: the database is queried as usual, but no data is returned.
In this case,
.I data_out
is ignored and may be NULL.
.RE

The return type of the data need not be the same as the type of the data stored
in the database.  Type conversion will be performed as necessary to return the
requested type.  If the
.I field_code
does not indicate a representation, but conversion from a complex value to a
purely real one is required, only the real portion of the requested vector will
be returned.

.SH RETURN VALUE
In all cases,
.BR getdata ()
returns the number of samples (not bytes) successfully read from the database.
If the end-of-field is encountered before the requested number of samples have
been read, a short count will result.  The library does not consider this an
error.  Requests for data before the beginning-of-field marker, which may have
been shifted from frame zero by the presence of a
.B FRAMEOFFSET
directive, will result in the the data being padded at the front by NaN or zero
depending on whether the return type is of floating point or integral type.

If an error has occurred, zero is returned and the dirfile error
will be set to a non-zero value.  Possible error values are:
.TP 8
.B GD_E_ALLOC
The library was unable to allocate memory.
.TP
.B GD_E_BAD_CODE
The field specified by
.IR field_code ,
or one of the fields it uses for input, was not found in the database.
.TP
.B GD_E_BAD_DIRFILE
An invalid
.I dirfile
was supplied.
.TP
.B GD_E_BAD_FIELD_TYPE
The supplied
.I field_code
referred to a 
.B CONST
or 
.B STRING
field.  The caller should use
.BR get_constant (3),
or
.BR get_string (3)
instead.
.TP
.B GD_E_BAD_REPR
The representation suffix specified in
.IR field_code ,
or in one of the field codes it uses for input, was invalid.
.TP
.B GD_E_BAD_SCALAR
A scalar field used in the definition of the field was not found, or was not of
scalar type.
.TP
.B GD_E_BAD_TYPE
An invalid
.I return_type
was specified.
.TP
.B GD_E_DIMENSION
A scalar field was found where a vector field was expected.
.TP
.B GD_E_INTERNAL_ERROR
An internal error occurred in the library while trying to perform the task.
This indicates a bug in the library.  Please report the incident to the
maintainer.
.TP
.B GD_E_OPEN_LINFILE
An error occurred while trying to read a LINTERP table from disk.
.TP
.B GD_E_RAW_IO
An error occurred while trying to open or read from a file on disk containing
a raw field.
.TP
.B GD_E_RECURSE_LEVEL
Too many levels of recursion were encountered while trying to resolve
.IR field_code .
This usually indicates a circular dependency in field specification in the
dirfile.
.TP
.B GD_E_UNKNOWN_ENCODING
The encoding scheme of a RAW field could not be determined.  This may also
indicate that the binary file associated with the RAW field could not be found.
.TP
.B GD_E_UNSUPPORTED
Reading from dirfiles with the encoding scheme of the specified dirfile is not
supported by the library.  See
.BR dirfile-encoding (5)
for details on dirfile encoding schemes.
.RE
.P
The dirfile error may be retrieved by calling
.BR get_error (3).
A descriptive error string for the last error encountered can be obtained from
a call to
.BR get_error_string (3).
.SH LIMITATIONS
The
.B PHASE
field type is poorly defined, since a forward-shifted
.B PHASE
field will always encounter the end-of-field marker before its input field does.
This has ramifications when using
.BR getdata ()
with streaming data.  The Dirfile
Standards make tacit admission to this problem by indicating the results of
reading a
.B PHASE
field beyond the beginning- or end-of-field is "implementation dependent" (see
.BR dirfile-format (5)).
As with any other field,
.BR getdata ()
will return a short count whenever a read from a
.B PHASE
field encounters the end-of-field marker.

Backward-shifted
.B PHASE
fields do not suffer from this problem, since
.BR getdata ()
pads reads past the beginning-of-field marker with NaN or zero as appropriate.
Database creators who wish to use the
.B PHASE
field type with streaming data are encouraged to work around this limitation
by only using backward-shifted 
.B PHASE
fields, by writing
.B RAW
data at the maximal time lag, and then back-shifting all data which should have
been written earlier.  Another possible work-around is to write
systematically less data to the first
.B RAW
field in proportion to the maximal forward phase shift.  This method will work
with applications which respect the database size reported by
.BR get_nframes (3)
resulting in these applications effectively ignoring all frames past the frame
containing the maximally forward-shifted
.B PHASE
field's end-of-field marker.

.SH SEE ALSO
.BR dirfile (5),
.BR dirfile-encoding (5),
.BR dirfile_open (3),
.BR get_constant (3),
.BR get_error (3),
.BR get_error_string (3),
.BR get_nframes (3),
.BR get_spf (3),
.BR get_string (3),
.BR putdata (3)


NEWS
New in version 0.6.1:

  API Changes

  * The getdata.h header no longer includes complex.h explicitly.

  Legacy API Changes

  * A potential segfault relating to error reporting in the legacy API has
    been fixed.
  
  Bindings

  * C++ BUG FIX: Several bugs preventing compilation of the C++ test-suite under
    Fedora 13 have been fixed.
  * IDL BUG FIX: Several segfaults in the IDL bindings have been eradicated.

  Miscellaneous

  * Libtool has been upgraded to 2.2.6b, which provides proper support for
    Fortran-9x.  The outdated, internal libltdl has been removed.  Building
    modules now requires an installed libltdl.

|==============================================================================|

New in version 0.6.0:

  License

  * The GetData library and its bindings are now distributed under the GNU
    Lesser Public License, version 2.1 or later.

  Dirfile Changes

  * GetData now support Dirfile Standards Version 7.  Standards Version 7 adds
    complex data types, a signed bitfield (SBIT), a polynomial derived field
    (POLYNOM), and other features detailed below.
  * Metafields can now be specified using a normal format specification line and
    their full (slashed) field code.
  * Field codes can now contain a "representation suffix", which allows
    specifying a complex norm used to convert complex valued data into purely
    real data.
  * The n_fields parameter in a LINCOM specification is now optional.
  * A new encoding scheme which supports LZMA compressed data using the .xz
    container format has been added.
  * Look-up tables may contain a complex y-value.  This prohibits writing to
    the associated LINTERP field.

  Library Changes

  * The field parameters spf (RAW), bitnum and numbits (BIT and SBIT) are now
    restricted to 16-bit numbers, which should be more than enough.  The public
    API uses gd_spf_t and gd_bit_t for these quantities.
  * The shift field parameter is now a 64-bit integer type, called gd_shift_t.
  * The dirfile flags, including encoding and endianness symbols are now
    explicitly long ints.  Previously these quantities mixed long int with int
    arbitrarily.  This affects the public API.
  * The parser callback has received a new, API breaking signature, which should
    hopefully reduce the amount of API breakage in the future.  It is also now
    possible for the caller to pass data through GetData to the callback, if
    desired.
  * Computation of derived fields can now handle complex valued inputs, and
    computation will occur in the complex plane when required.
  * The computation of LINCOM fields has been re-written and is now ~20% faster
    for LINCOMs with 2 or 3 input fields.  (No change for LINCOMs of 1 field.)
  * A new open flag, GD_PRETTY_PRINT, tells the library to attempt to write a
    nicer looking format file.  What this specifically means, isn't part of the
    public API and shouldn't be relied on in portable code.
  * BUG FIX: dirfile_madd_spec() and dirfile_add_spec() didn't properly flag
    the fragments they modified as dirty, leading to possible loss of metadata.
  * BUG FIX: calling get_nframes on a bzip2 encoded dirfile would report the
    wrong dirfile length.
  * BUG FIX: LINTERP tables are now closed after loading, resolving a file
    descriptor leak of great age (ie. pre-0.3.0).
  * BUG FIX: When re-writing a format file fragment containing fields with
    CONST scalar field parameters, GetData no longer replaces the CONST field
    codes in the field specification lines with the value of the CONST field.
  * BUG FIX: when moving a field with dirfile_move(), the fragment index in
    the entry objects of metafields defined for the field were not updated.
    (The metafields were placed in the correct fragment when the metadata was
    flushed to disk, despite this.)
  * BUG FIX: Calling getdata() or putdata() on 64-bit architectures in certain
    situations involving a non-zero frameoffset no longer results in bizarrity.
  * BUG FIX: A memory leak associated with RAW field metadata has been fixed.
  * BUG FIX: 64-bit integer literals in the format file are no longer truncated
    to double precision when read.

  API Changes

  * This release breaks ABI compatibility and, to a lesser extent, API
    compatibility.  The SOVERSION of the library has been incremented
    accordingly.
  * A C-89 API will be declared by getdata.h if GETDATA_C89_API is defined
    before inclusion.  If this symbol is not defined, the C API defined will
    include C-99 conforming complex types, and getdata.h will include complex.h
    if `complex' has not been defined before inclusion.
  * Due to the new, simple way of specifying META fields in Standards Version 7,
    dirfile_add_spec() and dirfile_alter_spec() can be used to modify
    metafields. dirfile_madd_spec() and dirfile_malter_spec() remain as well.
  * An array containing the CONST scalar field codes used when non-literal
    parameters are used in the specification of various different fields has
    been added to the gd_entry_t object.  As a side-effect of this,
    dirfile_add(), dirfile_alter_entry(), and dirfile_madd() can be used to
    manipulate non-literal field parameters.
  * A function, get_native_type(), now exists which returns the underlying data
    type of a specified field code.
  * A function, get_framenum(), is now available to perform a "reverse look-up"
    on a field: given a data value, it reports the frame number where the field
    has that value.  The field must be monotonic.
  * A function, dirfile_validate(), now exists which can be used to check the
    validity of a field code before use.
  * Two new error codes have been added: GD_E_DOMAIN, potentially returned by
    get_framenum(), and GD_E_BAD_REPR, indicating an unrecognised representation
    suffix, or an attempt to write to a representation.
  * A new syntax error suberror code has been added: GD_E_FORMAT_LITERAL, which

bindings/f77/getdata.f
C     (C) 2008 D. V. Wiebe
C
C     CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
C
C     This file is part of the GetData project.
C
C     This program is free software; you can redistribute it and/or
C     modify it under the terms of the GNU General Public License as
C     published by the Free Software Foundation; either version 2 of the
C     License, or (at your option) any later version.
C
C     GetData is distributed in the hope that it will be useful, but
C     WITHOUT ANY WARRANTY; without even the implied warranty of
C     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
C     General Public License for more details.
C
C     You should have received a copy of the GNU General Public License
C     along with GetData; if not, write to the Free Software Foundation,
C     Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
C

C     Fortran 77 parameters for getdata.  This file defines handy
C     constants useful to Fortran programs

C     Error codes
C     Corresponding to GD_E_OK
      INTEGER GD_EOK
      PARAMETER (GD_EOK=0)
C     Corresponding to GD_E_OPEN
      INTEGER GD_EOP
      PARAMETER (GD_EOP=1)
C     Corresponding to GD_E_FORMAT
      INTEGER GD_EFO
      PARAMETER (GD_EFO=2)
C     Corresponding to GD_E_TRUNC
      INTEGER GD_ETR
      PARAMETER (GD_ETR=3)
C     Corresponding to GD_E_CREAT
      INTEGER GD_ECR
      PARAMETER (GD_ECR=4)
C     Corresponding to GD_E_BAD_CODE
      INTEGER GD_EBC
      PARAMETER (GD_EBC=5)
C     Corresponding to GD_E_BAD_TYPE
      INTEGER GD_EBT
      PARAMETER (GD_EBT=6)
C     Corresponding to GD_E_RAW_IO
      INTEGER GD_ERW
      PARAMETER (GD_ERW=7)
C     Corresponding to GD_E_OPEN_INCLUDE
      INTEGER GD_EOI
      PARAMETER (GD_EOI=8)
C     Corresponding to GD_E_INTERNAL_ERROR
      INTEGER GD_EIE
      PARAMETER (GD_EIE=9)
C     Corresponding to GD_E_EMPTY
      INTEGER GD_EEM
      PARAMETER (GD_EEM=10)
C     Corresponding to GD_E_ALLOC
      INTEGER GD_EAL
      PARAMETER (GD_EAL=11)
C     Corresponding to GD_E_RANGE
      INTEGER GD_ERA
      PARAMETER (GD_ERA=12)
C     Corresponding to GD_E_OPEN_LINFILE
      INTEGER GD_EOL
      PARAMETER (GD_EOL=13)
C     Corresponding to GD_E_RECURSE_LEVEL
      INTEGER GD_ERL
      PARAMETER (GD_ERL=14)
C     Corresponding to GD_E_BAD_DIRFILE
      INTEGER GD_EBD
      PARAMETER (GD_EBD=15)
C     Corresponding to GD_E_BAD_PUT_FIELD
      INTEGER GD_EBP
      PARAMETER (GD_EBP=16)
C     Corresponding to GD_E_ACCMODE
      INTEGER GD_EAC
      PARAMETER (GD_EAC=17)
C Open flags
C     Correponding to GD_RDONLY
      INTEGER GD_RO
      PARAMETER (GD_RO=0)
C     Corresponding to GD_RDWR
      INTEGER GD_RW
      PARAMETER (GD_RW=2)
      INTEGER GD_CR
C     Corresponding to GD_CREAT
      PARAMETER (GD_CR=64)
      INTEGER GD_EX
C     Corresponding to GD_EXCL
      PARAMETER (GD_EX=128)
      INTEGER GD_TR
C     Corresponding to GD_TRUNC
      PARAMETER (GD_TR=512)
      INTEGER GD_BE
C     Corresponding to GD_BIG_ENDIAN
      PARAMETER (GD_BE=65536)
      INTEGER GD_LE
C     Corresponding to GD_LITTLE_ENDIAN
      PARAMETER (GD_LE=131072)
      INTEGER GD_FE
C     Corresponding to GD_FORCE_ENDIAN
      PARAMETER (GD_FE=262144)
      INTEGER GD_PE
C     Corresponding to GD_PEDANTIC
      PARAMETER (GD_PE=1048576)
C Data types -- F77 can't handle unsigned types, so we skip them
C     Corresponding to GD_NLL
      INTEGER GD_NLL
      PARAMETER (GD_NLL=0)
C     Corresponding to GD_INT8
      INTEGER GD_I8
      PARAMETER (GD_I8=17)
C     Corresponding to GD_INT16
      INTEGER GD_I16
      PARAMETER (GD_I16=18)
C     Corresponding to GD_INT32
      INTEGER GD_I32
      PARAMETER (GD_I32=20)
C     Corresponding to GD_INT64
      INTEGER GD_I64
      PARAMETER (GD_I64=24)
C     Corresponding to GD_FLOAT32
      INTEGER GD_F32
      PARAMETER (GD_F32=36)
C     Corresponding to GD_FLOAT64
      INTEGER GD_F64
      PARAMETER (GD_F64=40)

C Externals
C     Corresponding to dirfile_open(3)
      EXTERNAL GDFOPN
C     Corresponding to dirfile_close(3)
      EXTERNAL GDFCLS
C     Corresponding to getdata(3)
      EXTERNAL GDFGET
C     Corresponding to get_n_frames(3)
      EXTERNAL GDFNFR
C     Corresponding to get_samples_per_frame(3)
      EXTERNAL GDFSPF
C     Corresponding to putdata(3)
      EXTERNAL GDFPUT
C     Returns the dirfile->error member
      EXTERNAL GDFERR
C     Corresponding to getdata_error_string(3)
      EXTERNAL GDFSTR
